Integrate OpusTrainer

eu9ene opened the pull request:
- integrated [OpusTrainer](https://github.com/hplt-project/OpusTrainer) in train.sh
- added dataset importer that can augment datasets for evaluation
- removed teacher fine-tuning step. The pre-training and fine-tuning are now done in one step
- removed merge-augmented step
- adjusted pipeline settings to work with a higher amount of data
- modified the Snakemake pipeline accordingly but didn't test
- updated browsermt marian
- added docs
- added unit tests

Adjusting curriculum stages for the teacher is a bit tricky and might require further experimentation. I initially tried the 50-50 data mix which was probably too noisy because the teacher model stopped training too early, so I introduced a more complex schedule with the gradual adding of back-translations.
I also used `early-stopping: 30`.


Backward training run: https://firefox-ci-tc.services.mozilla.com/tasks/groups/G2FH3MaOTym3Vi9xKm_9zA
Increasing augmentation probs from 0.05 to 0.1 helped to increase performance for upper case sentences.

Completed teacher training run: https://firefox-ci-tc.services.mozilla.com/tasks/PsS4yUjVSjSZAY0kSe4JSQ/runs/0

Eval results for the teacher:

| dataset | augmentation | BLEU |
| --- | --- | --- |
| sacrebleu_wmt19 | no | 30.9 |
| sacrebleu_aug-upper_wmt19 | 10% upper case | 30.6 |
| sacrebleu_aug-upper-strict_wmt19 | all upper case | 28.4 |
| sacrebleu_aug-title-strict_wmt19 | all title case | 32 |
| sacrebleu_aug-title_wmt19 | 10% title case | 31 |

I reduced the amount of data for back translations to speed up testing.  In future, we can expand it and use `until backtranslated` instead of `until original` to train longer on back translations. This might improve quality. The quality of the current teacher is slightly lower than for the [OPUS one](https://opus.nlpl.eu/leaderboard/index.php?model=Tatoeba-MT-models%2Flit-eng%2FopusTCv20210807%2Bbt_transformer-big_2022-02-25&pkg=opusmt) (32.3). We could also try reducing the augmentation rate for title case.


Running on the full dataset takes a lot of time and is still in progress.

fixes #161 
fixes #209 

Training config:
```

datasets:
  # parallel training corpus
  train:
    - opus_CCAligned/v1
    - opus_DGT/v2019
    - opus_ECB/v1
    - opus_ECDC/v2016-03-16
    - opus_ELITR-ECA/v1
    - opus_ELRA-W0160/v1
    - opus_ELRC-2021-EUIPO_2017/v1
    - opus_ELRC-2717-EMEA/v1
    - opus_ELRC-2740-vaccination/v1
    - opus_ELRC-2878-EU_publications_medi/v1
    - opus_ELRC-3205-antibiotic/v1
    - opus_ELRC-3296-EUROPARL_covid/v1
    - opus_ELRC-3467-EC_EUROPA_covid/v1
    - opus_ELRC-3568-EUR_LEX_covid/v1
    - opus_ELRC-3609-presscorner_covid/v1
    - opus_ELRC-405-President_Lithuania/v1
    - opus_ELRC-425-Lithuanian_legislati/v1
    - opus_ELRC-4270-NTEU_TierA/v1
    - opus_ELRC-5067-SciPar/v1
    - opus_ELRC-590-www.lrs.lt/v1
    - opus_ELRC-591-www.lb.lt/v1
    - opus_ELRC-592-kam.lt/v1
    - opus_ELRC-EC_EUROPA/v1
    - opus_ELRC-EMEA/v1
    - opus_ELRC-EUIPO_2017/v1
    - opus_ELRC-EUROPARL_covid/v1
    - opus_ELRC-EUR_LEX/v1
    - opus_ELRC-EU_publications/v1
    - opus_ELRC-antibiotic/v1
    - opus_ELRC-presscorner_covid/v1
    - opus_ELRC-vaccination/v1
    - opus_ELRC-wikipedia_health/v1
    - opus_ELRC_2922/v1
    - opus_ELRC_2923/v1
    - opus_ELRC_3382/v1
    - opus_EMEA/v3
    - opus_EUbookshop/v2
    - opus_EUconst/v1
    - opus_Europarl/v8
    - opus_GNOME/v1
    - opus_JRC-Acquis/v3.0
    - opus_KDE4/v2
    - opus_NLLB/v1
    - opus_NeuLab-TedTalks/v1
    - opus_OpenSubtitles/v2018
    - opus_ParaCrawl/v9
    - opus_QED/v2.0a
    - opus_TED2020/v1
    - opus_Tatoeba/v2023-04-12
    - opus_TildeMODEL/v2018
    - opus_Ubuntu/v14.10
    - opus_WikiMatrix/v1
    - opus_XLEnt/v1.2
    - opus_bible-uedin/v1
    - opus_wikimedia/v20230407
    - mtdata_EU-eac_forms-1-eng-lit
    - mtdata_EU-eac_reference-1-eng-lit
    - mtdata_EU-ecdc-1-eng-lit
    - mtdata_Statmt-wiki_titles-1-lit-eng
  # datasets to merge for validation while training
  devtest:
    - flores_aug-mix_dev
    - sacrebleu_aug-mix_wmt19/dev
    - mtdata_aug-mix_Neulab-tedtalks_dev-1-eng-lit
  # datasets for evaluation
  test:
    - flores_devtest
    - flores_aug-mix_devtest
    - flores_aug-title_devtest
    - flores_aug-title-strict_devtest
    - flores_aug-upper_devtest
    - flores_aug-upper-strict_devtest
    - flores_aug-typos_devtest
    - sacrebleu_wmt19
    - sacrebleu_aug-mix_wmt19
    - sacrebleu_aug-title_wmt19
    - sacrebleu_aug-title-strict_wmt19
    - sacrebleu_aug-upper_wmt19
    - sacrebleu_aug-upper-strict_wmt19
    - sacrebleu_aug-typos_wmt19
    - mtdata_Neulab-tedtalks_test-1-eng-lit
    - mtdata_aug-mix_Neulab-tedtalks_test-1-eng-lit
  # monolingual datasets (ex. paracrawl-mono_paracrawl8, commoncrawl_wmt16, news-crawl_news.2020)
  # to be translated by the teacher model
  mono-src:
    - news-crawl_news.2022
    - news-crawl_news.2021
    - news-crawl_news.2020
    - news-crawl_news.2019
    - news-crawl_news.2018
   # to be translated by the backward model to augment teacher corpus with back-translations
  # leave empty to skip augmentation step (high resource languages)
  mono-trg:
    - news-crawl_news.2022
    - news-crawl_news.2021
    - news-crawl_news.2020
    - news-crawl_news.2019
    - news-crawl_news.2018
    - news-crawl_news.2017
    - news-crawl_news.2016
experiment:
  src: lt
  trg: en
  name: opustrainer
  vocab: NOT-YET-SUPPORTED
  bicleaner:
    default-threshold: 0.5
    dataset-thresholds:
      opus_CCAligned/v1: 0.7
      opus_OpenSubtitles/v2018: 0.8
      opus_ParaCrawl/v9: 0
      opus_WikiMatrix/v1: 0.7
      mtdata_Statmt-wiki_titles-1-lit-eng: 0.7
      opus_bible-uedin/v1: 0.7
  best-model: chrf
  split-length: 2000000
  backward-model: NOT-YET-SUPPORTED
  spm-sample-size: 10000000
  spm-vocab-size: 32000
  teacher-ensemble: 2
  mono-max-sentences-src: 500000000
  mono-max-sentences-trg: 500000000
  use-opuscleaner: 'false'
marian-args:
  decoding-teacher:
    precision: float16
    mini-batch-words: '4000'
  training-student:
    early-stopping: '20'
  decoding-backward:
    beam-size: '12'
    mini-batch-words: '2000'
  training-backward:
    early-stopping: '5'
  training-teacher:
    early-stopping: '30'
  training-student-finetuned:
    early-stopping: '20'
taskcluster:
  split-chunks: 10
target-stage: all


```

eu9ene commented:
I'm still testing it on the full dataset but it won't change much, so it's good for review now.

eu9ene commented:
It's interesting how it performs better for title case. We can probably reduce augmentation rate for it.

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+    run_cmd([comp_cmd, "-d", compressed_src])
+    run_cmd([comp_cmd, "-d", compressed_trg])
+
+    with open(uncompressed_src) as f:
```
Issue: This command is reading the entire file into memory, processing it, and writing it out. This means the dataset will be loaded completely into memory, rather than stream it.

Suggestion: How about creating an iterator that streams everything, something like this? (I didn't try running this, but just thought through the problem)

```py
class CorpusIter:
    def __init__(self, f_src_in, f_trg_in, f_src_out, f_trg_out):
        self.f_src = f_src
        self.f_trg = f_trg

    def __iter__(self):
        return self

    def __next__(self):
        src_sentence = self.f_src.readline()
        trg_sentence = self.f_trg.readline()
        return f"{src_sentence}\t{trg_sentence}"

with \
    open(uncompressed_src) as f_src_in, \
    open(uncompressed_trg) as f_trg_in, \
    open(uncompressed_src2, "w") as f_src_out, \
    open(uncompressed_trg2, "w") as f_trg_out:

    corpus = CorpusIter(f_src_in, f_trg_in, f_src_out, f_trg_out)

    for line in modifer(corpus):
        (modified_src, modified_trg) = line.split("\t")
        f_src_out.write(modified_src)
        f_src_out.write("\n")
        f_trg_out.write(modified_trg)
        f_trg_out.write("\n")

```

eu9ene replied:
We don't need to optimize this because it's used only for small evaluation datasets. We augment training datasets on the fly. I added a comment about it in [line 79](https://github.com/mozilla/firefox-translations-training/pull/219/files#diff-60c5881d16ade61511201a410163753c2ef9c76cc3189d272ffad40b1568ecadR79).

We would definitely need to stream for large datasets.

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+
+def main() -> None:
+    print(f"Running with arguments: {sys.argv}")
+    parser = argparse.ArgumentParser()
```
Suggestion: The docs can be shown at the CLI level which can be nice:

```
argparse.ArgumentParser(
    description=__doc__,
    # Preserves whitespace in the help text.
    formatter_class=argparse.RawTextHelpFormatter
)
```

eu9ene replied:
Good suggestion, but this script is not user facing

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+        print("Downloading mono dataset")
+        run_cmd([os.path.join(current_dir, "download-mono.sh"), dataset, output_prefix])
+    else:
+        raise ValueError(f"Invalid dataset type: {type}. Allowed values: mono, corpus")
```
Praise: Thanks for validating!

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--type", metavar="TYPE", type=str, help="Dataset type: mono or corpus")
+    parser.add_argument("--dataset", metavar="DATASET", type=str, help="Full dataset identifier")
```
Suggestion: Can you give some example identifiers? That will make it clearer for newer contributors. It's also not clear here how the augmentation identifier magic works.

eu9ene replied:
I provided examples [here](https://github.com/mozilla/firefox-translations-training/pull/219/files#diff-60c5881d16ade61511201a410163753c2ef9c76cc3189d272ffad40b1568ecadR150) and we have plenty documentation. I can add one examples here, so that it's more clear what this parameters is.

eu9ene replied:
added

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+    parser.add_argument("--type", metavar="TYPE", type=str, help="Dataset type: mono or corpus")
+    parser.add_argument("--dataset", metavar="DATASET", type=str, help="Full dataset identifier")
+    parser.add_argument(
+        "--output_prefix", metavar="OUTPUT_PREFIX", type=str, help="Write dataset here"
```
Suggestion: This is not clear to me what this is or means. It seems like it's where the data will be output.

eu9ene replied:
udpated

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+
+    if type == "corpus":
+        match = re.search(r"^(\w*)_(aug[a-z\-]*)?_?(.+)$", dataset)
```
Suggestion: Please give examples of how this identifier match is supposed to work. It's not clear from this identifier how it's supposed to work or what is valid. This also seems like it would be worth putting it into a utility if this is shared anywhere else in the code. Regexes like this can be a source of bugs, and it's not clear what the exact intent is for this one.

eu9ene replied:
added examples

gregtatum commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+        match = re.search(r"^(\w*)_(aug[a-z\-]*)?_?(.+)$", dataset)
+
+        if not match:
+            raise ValueError(f"Invalid dataset name: {dataset}")
```
Suggestion: Add a more helpful error message by defining _how_ to generate a valid dataset name. These names are somewhat magical, so it would be nice to be clear on how this works.

gregtatum reviewed (changes requested):
I didn't get all the way through the review, but I left some suggestions and spent most of my time on the `dataset_importer` script. I would suggest breaking things out a bit for easier review and landing. At this point the marian upgrade seems the easiest thing to break out at the very least as I think that would be an easy review merge.

I'm skipping review on the Snakemake files.

The dataset importer script could be split out into a separate PR, and landed separately. That would make it easier to get the reviews done. I left most of my feedback there, but haven't had time to look at other scripts.

gabrielBusta commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+            cmd,
+            universal_newlines=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
```
iirc that wasn't working for some reason, maybe opus trainer outputs errors to STDOUT?

eu9ene replied:
I had issues with this, I think OpusTrainer outputs something to stderr that's not visible in Taskcluster UI

gabrielBusta commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+        return batch
+
+
+MIX_PROB = 0.1  # 10% will be augmented in the mix
```
Would it make sense to pass this in?

eu9ene replied:
it's fine to leave it as a constant, we'll not change this often, maybe never

gabrielBusta commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+
+
+# we plan to use it only for small evaluation datasets
+def augment(output_prefix: str, aug_modifer: str):
```
This augment function is pretty long and handles multiple things (file handling, augmentation, compression). It might be good to split it into smaller functions.

gabrielBusta commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,90 @@ mkdir -p "${model_dir}/tmp"
+            <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${trg}.${ARTIFACT_EXT}") \
+            <(${COMPRESSION_CMD} -dc "${train_aln}") \
+            >"${tsv_dataset}"
+#            | pigz -c >"${tsv_dataset}"
```
If this is not needed, it's better to remove it.

eu9ene replied:
removed

gabrielBusta commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,90 @@ mkdir -p "${model_dir}/tmp"
-  --valid-log "${model_dir}/valid.log" \
-  "${extra_params[@]}"
+# Replace the path to vocab (required for alignments to work)
+sed -i -e "s#<vocab>#${vocab}#g" "${new_config}"
```
TIL -i behaves differently in MacOS...

eu9ene replied:
I ran this on MacOS and it worked. But I don't know if we even want this to be testable locally. It's just too hard to set up the environment, get all the input files etc.

gregtatum replied:
Thought: Breaking the problem down into smaller pieces will make parts of it testable so we have confidence on this type of code. Right now my confidence for correctness is pretty low, but I think the risk for merging here is pretty low as well.

I've run into plenty of issues myself where macOS's sed behavior is different from Linux.

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+
+
+[OpusTrainer](https://github.com/hplt-project/OpusTrainer) is a training tool developed by the HPLT project. 
+It feeds training data to Marian and provides ability to do useful manipulations with the data, 
```
```suggestion
It feeds training data to Marian and provides the ability to do useful manipulations with the data, 
```

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+
+## Data augmentation
+
+Data augmentation helps with making translation models more robust which is especially useful for usage with the noisy internet pages.
```
```suggestion
Data augmentation helps make translation models more robust, which is especially useful for usage with noisy internet pages.
```

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+
+Data augmentation helps with making translation models more robust which is especially useful for usage with the noisy internet pages.
+
+OpusTrainer augments data on the fly, which means it will generate unique data for each epoch of training. 
```
```suggestion
OpusTrainer augments data on the fly, meaning it will generate unique data for each epoch of training.
```

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+
+## Curriculum learning
+
+Ability to split training into multiples stages. Each stage is configurable to use a mix of different datasets.
```
```suggestion
Ability to split training into multiple stages. Each stage is configurable to use a mix of different datasets.
```

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+## Configuration
+
+OpusTrainer configuration files for the trained models are located in 
+[/pipeline/train/configs/opustrainer/](https://github.com/mozilla/firefox-translations-training/tree/main/pipeline/train/configs/opustrainer/) directory.
```
```suggestion
the [/pipeline/train/configs/opustrainer/](https://github.com/mozilla/firefox-translations-training/tree/main/pipeline/train/configs/opustrainer/) directory.
```

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+finetune:
+  - original 1.0
+  - until original inf # Fine-tuning only on original until the early stopping
+
```
```suggestion

```

gabrielBusta commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,130 @@
+
+### Supported modifiers
+
+`aug-typos` - applies typos with probability 0.1
```
```suggestion
`aug-typos` - applies typos with a probability of 0.1
```

gabrielBusta commented on pipeline/data/requirements/data.in:
```diff
@@ -0,0 +1,2 @@
+git+https://github.com/hplt-project/OpusTrainer.git@6c93fb667726504cad1cf5ab1aba140b99110f9d
```
Why are we using this revision? Why not [the PyPi package](https://pypi.org/project/opustrainer/)?

A comment on this file might help if someone wants to update it in the future and doesn't know the reason.

eu9ene replied:
it's the latest fixes from main that we need

eu9ene replied:
added comment

gabrielBusta commented on taskcluster/ci/train-backwards/kind.yml:
```diff
@@ -74,8 +75,11 @@ tasks:
                 - >-
+                    pip3 install --upgrade pip setuptools &&
+                    pip3 install -r $VCS_PATH/pipeline/train/requirements/train.txt &&
+                    export PATH="$HOME/.local/bin:$PATH" &&
```
What is in `.local/bin`?

bhearsum replied:
This is where pip installs to without a virtualenv, I believe.

eu9ene replied:
opustrainer-train is there

gabrielBusta commented on pipeline/train/requirements/train.in:
```diff
@@ -0,0 +1,2 @@
+git+https://github.com/hplt-project/OpusTrainer.git@9133e1525c7ee37f53ea14ee6a180152bf7ea192
```
Why use a different version of OpusTrainer here?

eu9ene replied:
should be the same, but they fixed a bug that affected only running marian (argument parsing), I can sync them

eu9ene replied:
synced

bhearsum commented on taskcluster/ci/finetune-teacher/kind.yml:
```diff
@@ -1,137 +0,0 @@
-# This Source Code Form is subject to the terms of the Mozilla Public
```
I'm curious why `finetune-teacher` is going away. (Is there a doc or something else written about this that I can refer to for the high level bits?)

eu9ene replied:
It's going away together with `merge-augmented` because now we can specify what datasets to mix and how to split training stages in the OpusTrainer config: https://github.com/mozilla/firefox-translations-training/pull/219/files#diff-efcaa59d90628026042cd6413c703108f907081e7e617bd363d653c6e263e7a1. This simplifies the pipeline a bit. You can look at https://github.com/hplt-project/OpusTrainer readme.

bhearsum commented on pipeline/train/train.sh:
```diff
@@ -12,13 +12,18 @@ model_type=$1
 training_type=$2
```
Not for now, but this is another script that _might_ benefit from being rewritten in Python (mostly because of all the fun argument handling). https://sh.readthedocs.io/en/latest/ might be useful if you choose to do that. (It's also by no means a requirement as far as I'm concerned.)

eu9ene replied:
Yeah, I had a dilemma about that, but I wanted to focus on making this work for now. I think we might give this a try when we'll be integrating the pre-trained OPUS-MT models because it will bring even more complexity.

eu9ene replied:
sh lib looks interesting. we should set up a pattern of how we're working with command line tools from Python

eu9ene commented on Snakefile:
```diff
@@ -90,15 +90,14 @@ biclean = f"{data_dir}/biclean"
 cache_dir = f"{data_dir}/cache"
```
It's fine, I didn't even test it. It works with dry-run though and I used it to regenerate the DAG graph. We want to minimize time on maintaining Snakemake.

bhearsum reviewed (approved):
I mostly focused on higher level bits, and as others have said it's a bit tough to review this in such a big batch. One thing that would help me understand it a bit better is to have a higher level explanation of the "why" behind removing certain steps, adding others, etc. (It's probably obvious to you, I realize, but I only have a very basic understanding of what OpusTrainer does, so it surprised me to see the removals of `merge-augmented` and `finetune-teacher`.)

eu9ene commented on Makefile:
```diff
@@ -109,7 +109,7 @@ dag:
 	  --profile=profiles/$(PROFILE) \
```
I had to update it in the end because the quantization broke. This might have something to do with the new vocab settings, not sure. I don't think we want to review the diff for the Marian update. We'll need to test the compatibility of the new models with our current WASM module just in case, but I don't think it can be broken. It's one line, so I don't know what would be the purpose of adding another PR for this.

eu9ene commented on configs/config.ci.yml:
```diff
@@ -19,50 +19,43 @@ experiment:
 
   best-model: chrf
 
-  use-opuscleaner: true
```
Good idea, ideally we want to move everything related to Snakemake to a separate directory. We can do that later.

eu9ene commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,90 @@ mkdir -p "${model_dir}/tmp"
+
+# Generate a new OpusTrainer config based on a template to fill paths of the datasets
+new_config="${model_dir}/config.opustrainer.yml"
+cp "configs/opustrainer/${model_type}.yml" "${new_config}"
```
Yes, it's required to train now.  I think there shouldn't be any downsides to using it. We need it for all training steps now because we always run data augmentation. If we wanted we could just disable augmentation and then it would behave similarly to Marian. It also brings the benefits of shuffling data without loading it in memory like Marian.

eu9ene commented on taskcluster/ci/dataset/kind.yml:
```diff
@@ -76,14 +81,20 @@ tasks:
-                    /builds/worker/artifacts/{dataset_sanitized}
+                    pip3 install --upgrade pip setuptools &&
+                    pip3 install -r $VCS_PATH/pipeline/data/requirements/data.txt &&
+                    chmod +x $VCS_PATH/pipeline/data/dataset_importer.py &&
```
right, I was testing something, removed

eu9ene commented on pipeline/train/configs/opustrainer/student.yml:
```diff
@@ -0,0 +1,25 @@
+# TODO: enable typos, issue https://github.com/hplt-project/OpusTrainer/issues/40
+#- Typos: 0.05
+# TODO: enable tags, currently doesn't work because of the issue with  tokenization
+# TODO: https://github.com/hplt-project/OpusTrainer/issues/38
```
replaced

eu9ene commented:
> I didn't get all the way through the review, but I left some suggestions and spent most of my time on the `dataset_importer` script. I would suggest breaking things out a bit for easier review and landing. At this point the marian upgrade seems the easiest thing to break out at the very least as I think that would be an easy review merge.
> 
> I'm skipping review on the Snakemake files.
> 
> The dataset importer script could be split out into a separate PR, and landed separately. That would make it easier to get the reviews done. I left most of my feedback there, but haven't had time to look at other scripts.

It's all very tied together and recommitting is a lot of work. I might give it a try next time but I doubt we'll have such significant changes often. Maybe only for integrating OPUS-MT pre-trained models but even then I don't know if we'll be able to split because it will be a big refactoring that touches a lot of places and should be merged together not to break the main branch. 

The nature of ML development is that it's all very experimental and you settle the code in the very end when the experiment is successful, so it's harder to determine the code parts that should be done in advance. ML-related PRs are more about landing a useful improvement to the model rather than just a useful piece of code on its own. For example, neither OpusTrainer integration nor evaluation script don't have much value on their own and become useful and landable only when it's proven that the experiment is successful and this change boosts the quality of the model and metrics which can't be done without a proper evaluation. And if experiment results are negative, this means the whole idea doesn't work and shouldn't be landed at all until it works. 

It's a bit of a paradigm shift from regular software development and the code is often more messy and experimental because what's important is metrics, performance and the flexibility to run more experiments more often. Bugs in code are usually very easy to detect with degrading metrics, so you have your training graphs instead of unit tests and CI (which can still be useful if the code is settled). Overall, it's more important to move faster with experimentation and improve metrics than to work on improving the code because models and metrics determine success of the project.

gregtatum reviewed (changes requested):
> It's all very tied together and recommitting is a lot of work. I might give it a try next time but I doubt we'll have such significant changes often. Maybe only for integrating OPUS-MT pre-trained models but even then I don't know if we'll be able to split because it will be a big refactoring that touches a lot of places and should be merged together not to break the main branch.

If you don't want to split it up, that is fine. I find it makes the review process harder, especially when there is disagreement on approach, as in now. It makes the process longer and messier to address in a big PR, compared to carving off parts that are not contentious to lower the review burden from one PR to a few. I'm happy to pair on splitting up work like this if you haven't had as much experience working this way, but it is up to you.

> The nature of ML development is that it's all very experimental and you settle the code in the very end when the experiment is successful, so it's harder to determine the code parts that should be done in advance. ML-related PRs are more about landing a useful improvement to the model rather than just a useful piece of code on its own. For example, neither OpusTrainer integration nor evaluation script don't have much value on their own and become useful and landable only when it's proven that the experiment is successful and this change boosts the quality of the model and metrics which can't be done without a proper evaluation. And if experiment results are negative, this means the whole idea doesn't work and shouldn't be landed at all until it works.

From this argument, it sounds like this work shouldn't even be merged at all. The main branch of projects should be tested and have a clean CI so that others can contribute to the code with confidence. If this code was just living on the Berlin cluster as a fork, it's one thing, but this is a shared project with multiple engineers working on it. I'm starting to scale up my own work here and would like to have confidence in what we are building. If you want to work on an experiment and try things, then I would suggest doing so in a branch, and when something is validated, then split out the pieces to merge it back in.

In this case, the `dataset_importer.py` file is a tool to that does not rely on experimentation to run. It can be validated in of itself with tests and automation. Once that tool is landed a branch could be used if you want to experiment with parts of the pipeline for other configurations. Otherwise we can do it all in this PR, but it might take a bit of time.

> It's a bit of a paradigm shift from regular software development and the code is often more messy and experimental because what's important is metrics, performance and the flexibility to run more experiments more often. Bugs in code are usually very easy to detect with degrading metrics, so you have your training graphs instead of unit tests and CI (which can still be useful if the code is settled). Overall, it's more important to move faster with experimentation and improve metrics than to work on improving the code because models and metrics determine success of the project.

I disagree with this. We can still build a robust platform that allows for experimentation. The goal here is to scale our work from a few languages to a hundred. In order to do that we need tools that we can rely on. In this case, for a script like `dataset_importer.py`, I would like to ensure the bare minimum of quality on it, and there are issues in the implementation. A test for something like that is not a very large request.

In order to move faster, we should be able to rely on the units of our code here and know that they are working. It's much better to have a 2 minute unit test run show that some of our code and assumptions are broken compared to debugging a 2 week training run that fails at a step later on for something that could easily be found with a test.

I would like to adjust expectations a bit on working as a team, in that with code review there is a bit more process in place, so things might not be an easy quick merge for something like this, but it is well worth the effort to share knowledge with the team on building things together.

This document has some good philosophical ideas on the whys of code review:
https://firefox-source-docs.mozilla.org/devtools/contributing/code-reviews.html

eu9ene commented:
@gregtatum all good thoughts, I'm sure we can find a balance between code quality and experimentation speed here. And I agree that basic quality checks are important, that's why I added CI. I prefer to leave this PR as is, since I don't have time to rewrite it to multiple parts and we already have some review history here.

I addressed all the comments, could you clarify what changes you are requesting?

eu9ene commented:
It's interesting why quantization is failing again, I updated browsermt marian because I thought it has something to do with the different vocab settings...

eu9ene commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+            universal_newlines=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            check=False,
```
To get the result and only then throw an exception so that we can print the result. I think it included more useful information when things fail.

eu9ene commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+    if aug_modifer not in modifier_map:
+        raise ValueError(f"Invalid modifier {aug_modifer}. Allowed values: {modifier_map.keys()}")
+
+    modifer = modifier_map[aug_modifer]
```
it's checked in line 84. if it's not in the map, it's an invalid modifier

eu9ene commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+            print(result.stdout)
+        raise
+
+    print(result.stdout)
```
Well ideally yes, but it has a very limited function here to run pigz/zstd, so it doesn't really matter.  I think it was easier to capture errors this way. I don't think it can cause memory issues.

eu9ene commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+    "aug-upper-strict": UpperCaseModifier(1.0),
+    "aug-mix": CompositeModifier(
+        [
+            # TODO: enable typos, issue https://github.com/hplt-project/OpusTrainer/issues/40
```
added issue

eu9ene commented on pipeline/data/dataset_importer.py:
```diff
@@ -0,0 +1,166 @@
+from opustrainer.types import Modifier
+
+
+class CompositeModifier:
```
added

eu9ene commented on envs/data.yml:
```diff
@@ -0,0 +1,16 @@
+name: bergamot-training
```
Let's do it later because it should also include moving other snakemake related things to a separated directory and making sure they still work.

eu9ene commented:
I missed a couple of comments yesterday, now I think I covered everything.

gregtatum commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,88 @@ mkdir -p "${model_dir}/tmp"
+      train_aln="${alns[index]}"
+      # alignments are currently used only for student that uses one stage of training
+      echo "### Generating tsv dataset with alignments ${alignments}"
+      paste <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${src}.${ARTIFACT_EXT}") \
```
Question: I'd be curious how performant this is, and what the memory model of it is. It's not clear to me how much this loads into memory, or how this works from an implementation perspective.

eu9ene replied:
This is a common pattern we use throughout the pipeline. I did not dig into implementation details but I didn't have issues with it.

gregtatum commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,88 @@ mkdir -p "${model_dir}/tmp"
+      rm "${train_aln}"
+      # when using tsv, marian requires --guided-alignments argument to be an index of the alignments in the tsv file
+      extra_params+=("--guided-alignment" "2")
+    else
```
Thought (follow-up): This logic would be good to test when we rewrite to python. It's quite hard to follow to know the correctness and performance characteristics of this. This area _is_ hot code, since it's the raw processing of datasets.

We could test and measure this outside of CI to make sure we are using these resources correctly.

gregtatum commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,129 @@
+It feeds training data to Marian and provides the ability to do useful manipulations with the data, 
+such as shuffling, mixing multiple datasets in the specified proportion, splitting training into multiple stages and augmentation.
+
+## Data augmentation
```
Praise: Thanks for the docs explaining the bigger picture here.

gregtatum commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,88 @@ mkdir -p "${model_dir}/tmp"
+IFS=',' read -ra datasets <<< "${train_set_prefixes}"
+IFS=',' read -ra alns <<< "${alignments}"
+# loop through the array and get both value and index
+for index in "${!datasets[@]}"; do
```
Question: Can you explain what this line and the next mean? My bash isn't good enough to understand these sections.

I also can't grok what exactly `IFS=',' read -ra datasets <<< "${train_set_prefixes}"` is doing here.

eu9ene replied:
It just expands the array to generate the indices to iterate on

eu9ene replied:
and the next one just element access by index

gregtatum reviewed (changes requested):
I got through the rest of the code. I spent a bit of time splitting it up locally to the relevant section so I could look through the related diffs. I left a bit more feedback, but most of it was questions or minor suggestions.

As we discussed in the meeting, I'll block on a simple test for the dataset_importer.py script that doesn't need to be hooked into CI.

Most of the relevant script changes were around 300 lines of code, and I spent most of my time there understanding the changes. Some of the `train.sh` code is not very understandable to me, but I think we can follow-up with converting it to python so that it's easier to reason with and understand the performance characteristics of the code.

gregtatum commented on tests/test_data_importer.py:
```diff
@@ -0,0 +1,116 @@
+        return f.readlines()
+
+
+def is_title_case(text):
```
Praise: Sweet solution!

gregtatum commented on tests/test_data_importer.py:
```diff
@@ -0,0 +1,116 @@
+    assert len(read_lines(output_trg)) > 0
+
+
+@pytest.mark.parametrize(
```
Thought: Huh, I didn't know about this decorator. Looks useful.

gregtatum commented on tests/test_data_importer.py:
```diff
@@ -0,0 +1,116 @@
+import gzip
```
Praise: Thank you for taking the time to write these tests.

gregtatum reviewed (approved):
Thanks for working through this review process with me, I appreciate it. Nice work here.

eu9ene commented on pipeline/train/train.sh:
```diff
@@ -30,42 +35,88 @@ mkdir -p "${model_dir}/tmp"
+for index in "${!datasets[@]}"; do
+    train_set_prefix="${datasets[index]}"
+    # OpusTrainer supports only tsv and gzip
+    # TODO: pigz is not found
```
added more details

eu9ene commented on taskcluster/ci/train-teacher/kind.yml:
```diff
@@ -82,42 +87,55 @@ tasks:
+            from-parameters: training_config.marian-args.training-teacher
         run:
             using: run-task
+            # order of comma separated datasets is important
```
Because they correspond to dataset0 and dataset1 which is documented in the docs

eu9ene commented on docs/opus-trainer.md:
````diff
@@ -0,0 +1,129 @@
+Example OpusTrainer config:
+```yaml
+datasets:
+  original: <dataset0> # Original parallel corpus
````
It says above:

```
`<dataset0>`, `<dataset1>` and `<vocab>` will be replaced by the training datasets passed in `pipeline/train/train.sh` script.
```
and then they are mixed based on the OpusTrainer config. 

I can add more, but which part is not clear?

gregtatum replied:
> will be replaced by the training datasets

This part is confusing to me. They will be replaced, but the _how_ and _why_ they are being replaced is not clear. If I remember from my review yesterday it's part of the bash script that I was finding confusing due to my lack of understanding of the bash/unix utilities.

eu9ene commented on docs/opus-trainer.md:
```diff
@@ -0,0 +1,129 @@
+
+### Supported modifiers
+
+`aug-typos` - applies typos with a probability of 0.1
```
added couple more examples

eu9ene commented on pipeline/train/spm-vocab.sh:
```diff
@@ -32,10 +32,17 @@ mkdir -p "${vocab_dir}"
+# byte_fallback - decomposes unknown pieces into UTF-8 bytes
+# user_defined_symbols - placeholders
+# character_coverage - CJK is recommended to have 0.9995, vocab languages probably you want 1
+"${MARIAN}/spm_train" --bos_id=-1 --eos_id=0 --unk_id=1 \
```
bos is -1

gregtatum replied:
Hah, whoops. I guess it is.
